pkg/kubelet/images/image_manager.go

```diff
@@ -84,52 +86,145 @@ func (m *imageManager) logIt(ref *v1.ObjectReference, eventtype, event, prefix,
 	}
 }
 
-// EnsureImageExists pulls the image for the specified pod and container, and returns
-// (imageRef, error message, error).
-func (m *imageManager) EnsureImageExists(pod *v1.Pod, container *v1.Container, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, string, error) {
-	logPrefix := fmt.Sprintf("%s/%s", pod.Name, container.Image)
-	ref, err := kubecontainer.GenerateContainerRef(pod, container)
-	if err != nil {
-		klog.Errorf("Couldn't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
+// getAsyncPullKey generates a key used to track async image pulls
+func getAsyncPullKey(pod *v1.Pod, container *v1.Container) string {
+	return fmt.Sprintf("%s_%s_%s", pod.UID, container.Name, container.Image)
+}
+
+// RemoveAsyncPullsForPod removes any tracked async pulls for a given pod
+func (m *imageManager) RemoveAsyncPullsForPod(pod *v1.Pod) {
+	for _, container := range pod.Spec.InitContainers {
+		asyncKey := getAsyncPullKey(pod, &container)
+		if _, exists := m.asyncPulls.Load(asyncKey); exists {
+			m.asyncPulls.Delete(asyncKey)
+		}
 	}
 
+	for _, container := range pod.Spec.Containers {
+		asyncKey := getAsyncPullKey(pod, &container)
+		if _, exists := m.asyncPulls.Load(asyncKey); exists {
+			m.asyncPulls.Delete(asyncKey)
+		}
+	}
+}
+
+// ensureImageExistsInternal has logic shared between EnsureImageExists and EnsureImageExistsAsync
+// (imageRef, error message, error, imageSpec, pullRequired).
+func (m *imageManager) ensureImageExistsInternal(ref *v1.ObjectReference, pod *v1.Pod, container *v1.Container, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, string, error, kubecontainer.ImageSpec, bool) {
```

Comment:
I wasn't entirely happy with the scale of the inputs and outputs on this method as it feels very bulky, however, after a few attempts at refactoring this felt like the best I could do to share the majority of code between the `Ensure` methods. Any thoughts on how to improve really welcome.